### ISSUE TITLE
Support custom finalizer in r/r_bg session

### ISSUE THREAD

**Issue by daattali** [r-lib/callr#148]
As far as I can tell, there is no way to tell the child r process to run some code when the process exits.

Suppose I spawn an R process that is meant to run in the background and perform some action every 5 seconds, and when the process dies I want to ensure it also runs that action. I've tried a few things such as the following but could not find a way to do this. 

```
test <- callr::r_bg(function() {
  exitenv <- new.env()
  reg.finalizer(exitenv, function(s) {
    write.csv(iris, "end.csv")
  }, onexit = TRUE)
  while(TRUE) {
    # do something
    Sys.sleep(1)
  }
})
test$kill()
```

I want the `end.csv` to be written when the process gets killed. Is there a way to do this?

**Comment by gaborcsardi** [comment]
Here is an example:

```r
test <- callr::r_bg(function() {
  tryCatch({
    Sys.sleep(4)
  }, finally = message("finalizing"))
})
Sys.sleep(1)
test$interrupt()
test$poll_io(1000)
test$read_error_lines()
```

Notes:
- The `Sys.sleep(1)` is needed, because if we interrupt immediately, then the R subprocess has not even started yet. There is not much we can do in this case, obviously the finalizers will not run if R is interrupted while starting up.
- `tryCatch(..., finally = expr)` makes sure that `expr` runs, no matter what, whether `...` exits cleanly or early via an error or interrupt. Your finalizer approach works as well, but this is cleaner I think.
- `test$interrupt()` sends and interrupt signal to the subprocess. This is like pressing `CTRL+C` or `ESC`. The background process may ignore this, though, e.g. if it is running C code. If you use `test$kill()`, that sends the SIGKILL signal to the subprocess, and it will be terminated immediately, and will not have the chance to run any finalizers. There is nothing you can do about this. If the subprocess is killed with `SIGKILL`, or crashes, then the finalizer cannot run.
- `test$poll_io()` makes sure that there is something to read from the subprocess. You don't really need this if you don't want to read anything.

So in summary, your code is actually good :), but use `$interrupt()` instead of `$kill()` for testing it.

**Comment by daattali** [comment]
Ah, of course, `interrupt()` is what I should be calling rather than `kill()`.

This does work when the process is explicitly interrupted. But it doesn't seem to do much when the R session dies of other causes. For example by closing RStudio or simply restarting the parent R session.

I'm adding the infinite loop back to my example because the point is that I want the process to be long lived, until something external interrupts it. I'm also writing a file because it's easy to see the output.

```
test <- callr::r_bg(function() {
  tryCatch({
    while(TRUE) {
      Sys.sleep(1)
    }
  }, finally = write.csv(iris, "end.csv"))
})
Sys.sleep(1)
test$interrupt()
```

The above works. But if you replace `test$interrupt()` with `rstudioapi::restartSession()` then the output file doesn't get written.

**Comment by gaborcsardi** [comment]
Here is an idea. You can start the process with `cleanup = FALSE`, and put a finalizer on it, that calls `$interrupt()` first, and then possibly `$kill()` if the interrupt does not work. E.g.

```r
test <- callr::r_bg(cleanup = FALSE, function() {
  tryCatch({
    while(TRUE) {
      Sys.sleep(1)
    }  
  }, finally = write.csv(iris, "end.csv"))
})
reg.finalizer(
  test, 
  function(p) { p$interrupt(); p$poll_io(1000); p$kill() },
  onexit = TRUE
)
```

I am not entirely sure what happens if the main process gets killed or crashes, because then it cannot call the finalizer. It seems that on Unix at least, the subprocess keeps running happily...

**Comment by daattali** [comment]
I just found another workaround, a minute before you posted this.

Thanks for investing your time. Do you think there's a valid case to be made to add support for a finalizer parameter to `r()`/`r_bg()`? I won't be offended if you say it's too esoteric

**Comment by gaborcsardi** [comment]
Yeah, we could do that, it would be a nice feature. But to make it work we would also need to use `$interrupt()` in the finalizer of the processx object in the main process, to give a chance to the subprocess to clean up. But yeah, that is possible as well, probably with some configurable timeout.

**Comment by daattali** [comment]
Actually looks like my workarounds don't work. Will be awesome if a `finalize` parameter is added :)

**Comment by gaborcsardi** [comment]
How about https://github.com/r-lib/callr/issues/148#issuecomment-602121465? Is that not good?

**Comment by daattali** [comment]
It doesn't work when the , see reprex below. Ending the parent R session causes the test session to end (which is visible by the fact that files are no longer being generated every second), but the final `end` file isn't getting created.

```
test <- callr::r_bg(cleanup = FALSE, function() {
  tryCatch({
    while(TRUE) {
      file.create(as.character(as.integer(Sys.time())))
      Sys.sleep(1)
    }  
  }, finally = file.create("end"))
})
reg.finalizer(
  test, 
  function(p) { p$interrupt(); p$poll_io(1000); p$kill() },
  onexit = TRUE
)

Sys.sleep(5)
rstudioapi::restartSession()
```

**Comment by gaborcsardi** [comment]
Interesting. This does work fine for me on macOS and RStudio 1.3.778.

**Comment by daattali** [comment]
You're right, I tried it on linux and it worked. This seems to be a windows specific issue

**Comment by gaborcsardi** [comment]
This is pretty hard to do reliably, considering that the child process may block the interrupt, so I decided that I don't implement it now.
